src/sparsebase/utils/io/reader.cc

```diff
@@ -125,14 +126,223 @@ EdgeListReader<IDType, NNZType, ValueType>::~EdgeListReader(){};
+  } else if (field == "complex") {
+    options.field =
+        MTXReader<IDType, NNZType, ValueType>::MTXFieldOptions::complex;
+    if constexpr(std::is_same<void, ValueType>::value) throw ReaderException("You are reading the values of the matrix market file into a void array");
+  } else if (field == "integer") {
+    options.field =
+        MTXReader<IDType, NNZType, ValueType>::MTXFieldOptions::integer;
+    if constexpr(std::is_same<void, ValueType>::value) throw ReaderException("You are reading the values of the matrix market file into a void array");
+  } else if (field == "pattern") {
+    options.field =
+        MTXReader<IDType, NNZType, ValueType>::MTXFieldOptions::pattern;
+  } else {
+    throw ReaderException(
+        "Illegal value for the 'field' option in matrix market header");
+  }
+  // parsing symmetry
+  if (symmetry == "general") {
+    options.symmetry =
+        MTXReader<IDType, NNZType, ValueType>::MTXSymmetryOptions::general;
+  } else if (symmetry == "symmetric") {
+    options.symmetry =
+        MTXReader<IDType, NNZType, ValueType>::MTXSymmetryOptions::symmetric;
+  } else if (symmetry == "skew-symmetric") {
+    options.symmetry = MTXReader<IDType, NNZType,
+                                 ValueType>::MTXSymmetryOptions::skew_symmetric;
+  } else if (symmetry == "hermitian") {
+    options.symmetry =
+        MTXReader<IDType, NNZType, ValueType>::MTXSymmetryOptions::hermitian;
+    throw ReaderException("Matrix market reader does not currently support hermitian symmetry.");
+  } else {
+    throw ReaderException(
+        "Illegal value for the 'symmetry' option in matrix market header");
+  }
+  return options;
+}
 
+template <typename IDType, typename NNZType, typename ValueType>
+template <bool weighted>
+format::COO<IDType, NNZType, ValueType> *
+MTXReader<IDType, NNZType, ValueType>::ReadArrayIntoCOO() const {
+  std::ifstream fin(filename_);
+  // Ignore headers and comments:
+  while (fin.peek() == '%')
+    fin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+  // Declare variables: (check the types here)
+
+  format::DimensionType M, N;
+
+  fin >> M >> N;
+
+  format::DimensionType total_values = M * N;
+  IDType *long_rows = new IDType[total_values];
+  IDType *long_cols = new IDType[total_values];
+  ValueType *long_vals = nullptr;
+  if (!weighted)
+    long_vals= new ValueType[total_values];
+  NNZType num_nnz = 0;
+  for (format::DimensionType l = 0; l < total_values; l++) {
+    ValueType w;
+    fin >> w;
+
+    if (w != 0){
+      long_cols[num_nnz] = l/N;
+      long_rows[num_nnz] = l%N;
+      if (weighted)
+        long_vals[num_nnz] = w;
+      num_nnz++;
+
+    }
+    //nnz_counter += w != 0;
+  }
+
+  IDType *row = new IDType[num_nnz];
+  IDType *col = new IDType[num_nnz];
+  memcpy(row, long_rows, num_nnz*sizeof(IDType));
```

Comment:
This is a question: why did you prefer to use memcpy instead of std::copy?

Reply:
Honestly force of habit. But it seems [I should've used std::copy.](https://stackoverflow.com/questions/4707012/is-it-better-to-use-stdmemcpy-or-stdcopy-in-terms-to-performance) I'll make the change tonight and push again.